Fix generator factory result hint

stkrp opened the pull request:
Instead `YieldType` `SendType` was used.

> A generator can be annotated using the generic type [`Generator[YieldType, SendType, ReturnType]`](https://docs.python.org/3.14/library/collections.abc.html#collections.abc.Generator).

https://docs.python.org/3.14/library/typing.html#annotating-generators-and-coroutines

### Demo

```python
from collections.abc import Generator

from dishka import make_container, Provider, Scope


log = []

# Mypy: error: Incompatible types in "yield" 
#       (actual type "int", expected type "None")
# def generator_factory() -> Generator[None, int, None]:
#
# Mypy: no issues found
# def generator_factory() -> Generator[int, None, None]:
def generator_factory() -> Generator[int]:
    log.append("Enter")
    yield 1
    log.append("Exit")


provider = Provider(scope=Scope.APP)
provider.provide(generator_factory)
container = make_container(provider)

assert container.get(int) == 1
container.close()
assert log == ["Enter", "Exit"]
```

#### `Generator[int, None, None]`

```python
Traceback (most recent call last):
  File "demo.py", line 24, in <module>
    assert container.get(int) == 1
           ~~~~~~~~~~~~~^^^^^
  File ".venv/lib/python3.13/site-packages/dishka/container.py", line 167, in get
    return self._get_unlocked(key)
           ~~~~~~~~~~~~~~~~~~^^^^^
  File ".venv/lib/python3.13/site-packages/dishka/container.py", line 198, in _get_unlocked
    return self.parent_container._get(key)  # noqa: SLF001
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^
  File ".venv/lib/python3.13/site-packages/dishka/container.py", line 177, in _get
    return self._get_unlocked(key)
           ~~~~~~~~~~~~~~~~~~^^^^^
  File ".venv/lib/python3.13/site-packages/dishka/container.py", line 192, in _get_unlocked
    raise NoFactoryError(
    ...<3 lines>...
    )
dishka.exceptions.NoFactoryError: Cannot find factory for (int, component='', scope=Scope.APP). Check scopes in your providers. It is missing or has invalid scope.
```

#### `Generator[int]`

```python
Traceback (most recent call last):
  File "demo.py", line 21, in <module>
    provider.provide(generator_factory)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.13/site-packages/dishka/provider/provider.py", line 134, in provide
    composite = provide_on_instance(
        source=source,
    ...<4 lines>...
        override=override,
    )
  File ".venv/lib/python3.13/site-packages/dishka/provider/make_factory.py", line 570, in provide_on_instance
    return _provide(
        provides=provides, scope=scope, source=source, cache=cache,
        is_in_class=False,
        recursive=recursive, override=override,
    )
  File ".venv/lib/python3.13/site-packages/dishka/provider/make_factory.py", line 534, in _provide
    factory = make_factory(
        provides=provides, scope=scope,
    ...<2 lines>...
        override=override,
    )
  File ".venv/lib/python3.13/site-packages/dishka/provider/make_factory.py", line 484, in make_factory
    return _make_factory_by_function(
        provides=provides,
    ...<5 lines>...
        check_self_name=True,
    )
  File ".venv/lib/python3.13/site-packages/dishka/provider/make_factory.py", line 337, in _make_factory_by_function
    provides = _clean_result_hint(factory_type, possible_dependency)
  File ".venv/lib/python3.13/site-packages/dishka/provider/make_factory.py", line 225, in _clean_result_hint
    possible_dependency = _generator_result(possible_dependency)
  File ".venv/lib/python3.13/site-packages/dishka/provider/make_factory.py", line 189, in _generator_result
    return get_args(hint)[1]
           ~~~~~~~~~~~~~~^^^
IndexError: tuple index out of range
```

### Environment

* dishka 1.7.2
* Python 3.13.7 (v3.13.7:bcee1c32211, Aug 14 2025, 19:10:51) [Clang 16.0.0 (clang-1600.0.26.6)]
* macOS 15.7.1 (24G231)

Tishka17 commented:
Thank you for your PR, it is definitely a bug. Please, provide the minimal fix for it (I suppose only index change is required). 

Also, Generator[int] is available on python 3.13+ only, so we need to setup conditional tests here

stkrp commented:
> Please, provide the minimal fix for it (I suppose only index change is required).

Ok, I'll remove that generalisation. I'd be grateful if you could tell me why this would be better, so it doesn't happen again in the future.

> Also, typing.Generator[int] is available on python 3.13+ only, so we need to setup conditional tests.

You're suggesting to add a similar pair of factories with deprecated `typing.Generator` (where unary `typing.Generator[int]` should be conditional), right? If so, I will add these test cases. I already did `HAS_PY_313` and `HAS_TV_DEFAULT` in the first version, but I deleted it because everything worked without it (`GenericAlias` supports any number of args):

```python
>>> import types
>>> types.GenericAlias(dict, (list, set, tuple, int, Exception, object))
dict[list, set, tuple, int, Exception, object]
```

stkrp commented:
Thanks for the clarification.

I have reverted generalisation and added tests for `typing.(Async)Generator`. To avoid creating false stubs, the condition had to be added not only in factories module, but also in tests module. This allows us to run tests only where they are really needed, and only on real objects.

**P.S.** I didn't add extra support for the short form of `Generator` to  `UnsupportedGeneratorReturnTypeError` building:
```python
# dishka.provider.make_factory._generator_result
elif origin is AsyncGenerator:
    args = ", ".join(_type_repr(a) for a in get_args(hint)) + ", None"
```
Formally, this is correct, so I decided to keep it simple.
